### PR TITLE
Pre-install to validate single mesh

### PR DESCRIPTION
**Pull request by nshankar13**
**Description**: Add a pre-install job for validate-single-mesh, which verifies that osm isn't already installed in the osm-system or kube-system namespaces. 

First, the job searches for the osm-controller pod in the "kube-system" namespace, which is where the AKS add-on deploys osm, and checks if either osm-arc (being installed) or the AKS-addon enforces single mesh. 

In the second part of the job, it searches through all the helm releases and sees if any of the chart names (not release names) contain "osm" or "osm-arc". If osm-arc already exists on the cluster, the job enforces that the .Release.Namespace of the current installation matches the namespace that osm-arc is already installed in. If osm is already installed on the cluster, then the job checks if either osm-arc (being installed) or the existing osm deployment enforces single mesh. 

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [X]
- Networking             [ ]
- Metrics                [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

File: charts/osm-arc/values.yaml [hunk 1]
```diff
@@ -9,6 +9,7 @@ Azure:
 
 OpenServiceMesh:
     ignoreNamespaces: "kube-system azure-arc arc-osm-system"
+    validateNamespaces: "kube-system osm-system"
```
Comment (SanyaKochhar) [inline charts/osm-arc/values.yaml:12]: Could this key be named something a bit more intuitive? I feel like outside of the `validate-single-mesh` spec I wouldn't know what it is

Comment (nshankar13) [reply charts/osm-arc/values.yaml]: I actually don't need this field anymore so I'm getting rid of it.

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        - name: validate-single-mesh-oss
+          image: {{ .Values.helm.image.name }}:{{ .Values.helm.image.tag }}
+          imagePullPolicy: IfNotPresent
+          env: 
+            - name: enforcesinglemesh
+              value: {{ .Values.osm.OpenServiceMesh.enforceSingleMesh | quote }}
+            - name: releasenamespace
+              value: {{ .Release.Namespace }}
+          command: 
+            - "/bin/sh"
```
Comment (draychev) [inline charts/osm-arc/templates/validate-single-mesh.yml:117]: @nshankar13 I worry about having to maintain one more bash script, which could make or break an OSM installation.
We should have similar functionality in the OSM CLI (go code).
Do you think it is reasonable to replace the `bash` here with a container, which has the same checks implemented in Go?

I'd propose we
 - [ ] containerize the OSM CLI
 - [ ] publish the new container
 - [ ] use said container here with the right (perhaps new) OSM CLI command to check for this

Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:117]: @draychev what exactly do you mean by make or break an OSM installation? Provided that the AKS add-on is always in the kube-system namespace, and the upstream/arc chart names remain as osm/osm-arc, I can't foresee any subsequent updates to the pre-install validation job.

Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:117]: Also, the validation for arc's enforce-single-mesh works a bit differently than how upstream does it, since for osm-arc we want to allow re-installation in the same namespace to update the version (see lines 136 - 139 in validate-single-mesh.yml in this PR).

Comment (ksubrmnn) [reply charts/osm-arc/templates/validate-single-mesh.yml:117]: @draychev the OSM CLI checks this on install. I've actually wondered if we should be moving this logic to the osm controller itself with some sort of webhook so that it would be uniform for upstream and downstream

OSM CLI Code: https://github.com/openservicemesh/osm/blob/main/cmd/cli/install.go#L217-L230

File: charts/osm-arc/templates/osm-label.yml [hunk 3]
```diff
@@ -8,7 +8,7 @@ metadata:
   name: osm-label-account
   namespace: {{ .Release.Namespace }}
   annotations:
-    "helm.sh/hook-weight": "20"
+    "helm.sh/hook-weight": "25"
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/osm-label.yml:11]: what's the motivation behind these changes?

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      terminationGracePeriodSeconds: 0
+      containers:
+        - name: validate-single-mesh-aks-addon
+          image: {{ .Values.kubectl.image.name }}:{{ .Values.kubectl.image.tag }}
+          imagePullPolicy: IfNotPresent
+          env: 
+            - name: enforcesinglemesh
+              value: {{ .Values.osm.OpenServiceMesh.enforceSingleMesh | quote }}
+            - name: validatenamespaces
+              value: {{ .Values.OpenServiceMesh.validateNamespaces }}
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/validate-single-mesh.yml:84]: What is this validateNamespaces field? How is it set?

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          command: 
+            - "/bin/sh"
+            - "-ec"
+            - |
+              osm_namespaces=$(kubectl get deployments -l app=osm-controller -A -o=jsonpath={..metadata.namespace})
+              for namespace in $osm_namespaces
+              do
+                for validateNamespace in ${validatenamespaces} 
+                do
+                  if [ "$namespace" = "kube-system" ]; then
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/validate-single-mesh.yml:94]: Could this aks logic be bundled with the oss logic? Since the only differentiator here is that aks installs in kube-system?

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            - |
+              apk add --no-cache bash
+              bash
+              apk add jq
+              i=0
+              helm_list=$(helm ls --all --all-namespaces -o json | jq)
+              chart=$(echo $helm_list | jq .[$i])
+              while [[ "$chart" != "null" ]]
+              do
+                chart_name=$(echo $chart | jq .chart)
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/validate-single-mesh.yml:128]: Why are you using helm here instead of getting the deployments?

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                chart_namespace=$(echo $chart | jq .namespace)
+                temp="${chart_name%\"}"
+                chart_name="${temp#\"}"
+                temp="${chart_namespace%\"}"
+                chart_namespace="${temp#\"}"
+                if [[ $(expr match "$chart_name" 'osm') != 0 ]]; then
+                  if [[ $(expr match "$chart_name" 'osm-arc') != 0 ]]; then
+                    if [[ "$chart_namespace" != "${releasenamespace}" ]]; then
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/validate-single-mesh.yml:136]: Can you add a comment for this logic saying that it is necessary for upgrades?

**Review by nshankar13 — COMMENTED** [review]
.

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 4 above)
Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:84]: Meant to delete this, not necessary.

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 5 above)
Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:94]: With the AKS add-on, you cannot see the helm release, as I believe the add-on is not installed via helm like OSS and Arc are. Thus the AKS-addon scenario needs to be handled differently from the OSS logic, where we need to use helm release data to distinguish between OSS and osm-arc releases (explained below). I thought it would make sense to do the OSS job in a separate container with the alpine-helm container image.

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 6 above)
Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:128]: Here, we need to use helm release name to distinguish between osm-arc and OSS releases, as OSS could be in any namespace.

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 5 above)
Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:94]: Actually, I think there is a way to do this kubectl without needing to use a separate container with helm. Will explore this.

Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:94]: Ok, I remember why I decided to do this via helm CLI. For OSS, the kubect metadata via `kubectl get deployment -l app=osm-controller -o json` doesn't include the "enforceSingleMesh" field, so the only way to access it is by parsing the values via `helm get values`.

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 6 above)
Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml:128]: Also, I don't think it is possible to access the enforceSingleMesh field for OSS without helm cli - parsing the metadata via kubectl wasn't working for me (but it does work for AKS addon and Arc).

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 4 above)
Comment (ksubrmnn) [reply charts/osm-arc/templates/validate-single-mesh.yml:84]: can you delete this and references to it (line 92)?

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 5 above)
Comment (ksubrmnn) [reply charts/osm-arc/templates/validate-single-mesh.yml:94]: https://github.com/openservicemesh/osm/blob/dde84836e54eef62aaf7688b9f2d1c4dd730d366/charts/osm/templates/osm-deployment.yaml#L10

The enforceSingleMesh is added as a label, I think you can use this with the `-l` flag?

Comment (nshankar13) [reply charts/osm-arc/templates/validate-single-mesh.yml]: Found another issue with trying to use kubectl, separate from enforceSingleMesh. 

Kubectl is able to parse the helm release name, but not the actual chart name. The chart name is the same across releases (either osm or osm-arc), thus this is what I use to distinguish between osm-arc and osm releases. I believe this can only be accessed via the helm CLI.

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            - name: enforcesinglemesh
+              value: {{ .Values.osm.OpenServiceMesh.enforceSingleMesh | quote }}
+          command: 
+            - "/bin/sh"
+            - "-ec"
+            - |
+              osm_namespaces=$(kubectl get deployments -l app=osm-controller -A -o=jsonpath={..metadata.namespace})
+              for namespace in $osm_namespaces
+              do
+                if [ "$namespace" = "kube-system" ]; then
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/validate-single-mesh.yml:90]: Why do we have this check for kube-system? We shouldn't install the mesh if enforceSingleMesh is true and any mesh exists in any namespace, right?

File: charts/osm-arc/templates/validate-single-mesh.yml [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                  if [ "${enforcesinglemesh}" = "true" ]; then
+                    echo "osm already exists in namespace $namespace. osm-arc enforces single-mesh cluster."
+                    exit 1
+                  else
```
Comment (ksubrmnn) [inline charts/osm-arc/templates/validate-single-mesh.yml:94]: I don't think this block needs to be wrapped in a conditional. It's fine that it is, but technically this logic applies whether or not enforceSingleMesh is true for the new install.

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 8 above)
Comment (ksubrmnn) [reply charts/osm-arc/templates/validate-single-mesh.yml:90]: Oh this is the workaround for upgrades for Arc right now. We know Arc will never be in kube-system and AKS will always be in kube-system. This isn't ideal, but hopefully when things change in the future, we can make this more robust

File: charts/osm-arc/templates/validate-single-mesh.yml (hunk 6 above)
Comment (ksubrmnn) [reply charts/osm-arc/templates/validate-single-mesh.yml:123]: as discussed offline, the label is only added if enforceSingleMesh is true

**Review by ksubrmnn — APPROVED** [review]

File: .pipelines/upgrade-e2e-job.yaml [hunk 10]
```diff
@@ -10,9 +10,9 @@ pr:
   branches:
     include:
     - main
-  paths:
-    include:
-      - charts/osm-arc/Chart.yaml
+  #paths:
+    #include:
+      #- charts/osm-arc/Chart.yaml
```
Comment (SanyaKochhar) [inline .pipelines/upgrade-e2e-job.yaml:15]: why dis commented out

**Review by SanyaKochhar — APPROVED** [review]

File: .pipelines/upgrade-e2e-job.yaml (hunk 10 above)
Comment (nshankar13) [reply .pipelines/upgrade-e2e-job.yaml:15]: Will recomment before merging, just wanted to run the upgrade job in this PR to ensure it works with pre-install validation.
